src/trajectory.jl

```diff
@@ -590,42 +608,30 @@ function isterminated(h::Hamiltonian, t::BinaryTree{<:ClassicNoUTurn})
 end
 
 """
-    isterminated(h::Hamiltonian, t::BinaryTree{<:GeneralisedNoUTurn})
-
+$(SIGNATURES)
 Detect U turn for two phase points (`zleft` and `zright`) under given Hamiltonian `h`
 using the generalised no-U-turn criterion.
 
 Ref: https://arxiv.org/abs/1701.02434
 """
-function isterminated(h::Hamiltonian, t::BinaryTree{<:GeneralisedNoUTurn})
-    rho = t.c.rho
+function isterminated(::GeneralisedNoUTurn, h::Hamiltonian, t::BinaryTree)
+    rho = t.ts.rho
     s = generalised_uturn_criterion(rho, ∂H∂r(h, t.zleft.r), ∂H∂r(h, t.zright.r))
     return Termination(s, false)
 end
 
 """
-    isterminated(
-        h::Hamiltonian, t::T, tleft::T, tright::T
-    ) where {T<:BinaryTree{<:StrictGeneralisedNoUTurn}}
-
+$(SIGNATURES)
 Detect U turn for two phase points (`zleft` and `zright`) under given Hamiltonian `h`
 using the generalised no-U-turn criterion with additional U-turn checks.
 
 Ref: https://arxiv.org/abs/1701.02434 https://github.com/stan-dev/stan/pull/2800
 """
 function isterminated(
-    h::Hamiltonian, t::T, tleft::T, tright::T
-) where {T<:BinaryTree{<:StrictGeneralisedNoUTurn}}
-    # Classic generalised U-turn check
-    t_generalised = BinaryTree(
-        t.zleft,
-        t.zright,
-        GeneralisedNoUTurn(t.c.rho),
-        t.sum_α,
-        t.nα,
-        t.ΔH_max
-    )
-    s1 = isterminated(h, t_generalised)
+    tc::StrictGeneralisedNoUTurn, h::Hamiltonian, t, tleft, tright
+)
+    # (Non-strict) generalised U-turn check
+    s1 = isterminated(GeneralisedNoUTurn(tc), h, t)
```

Comment:
@yebai Something less elegent would appear here, if we were to put `tc` as a field of `BinaryTree`.
That is, we have to manually reconstruct the tree, as what we are doing master ATM: https://github.com/TuringLang/AdvancedHMC.jl/blob/master/src/trajectory.jl#L620-L627

PS, you cannot do this with reconstruct, because the type changes, by which I mean
```julia
julia> using Parameters

julia> struct A{T}
         v::T
       end

julia> a = A(1.0)
A{Float64}(1.0)

julia> reconstruct(a, v="1")
ERROR: MethodError: Cannot `convert` an object of type String to an object of type Float64
```

I guess the root reason here is that, `tc` is indeed _not_ a property of a tree but a surely a property of `NUTS`: `isterminated` should take both the tree and the criterion as input, to decide if the termination happens. For the same tree, based on different termianiton criteria, it could terminate or not.

Reply:
I hope to keep the current design if possible. I think pure multiple distapching without parametric types is also Julian :)

Reply:
thanks for the clarification, I agree that the new design is more modular. However, I feel it introduces a bit more clutter as well. It would be nice to simplify a bit here. For example, we can keep the `turn_statistic` field's type generic. Using your illustration, this would look like 

```julia
julia> struct A
         v
       end

julia> a = A(1.0)
A{Float64}(1.0)

julia> reconstruct(a, v="1")

```

Then, we can modify the function `isterminated` to dispatch based on the type of `BinaryTree.turn_statistic` rather than the type of `BinaryTree`.

Reply:
Making it generic will cause type instability issues whenever `a.v` is used for dispatching functions.

For example,
```julia
julia> struct A
       v
       end

julia> f(::AbstractFloat) = 1.0
f (generic function with 1 method)

julia> f(::AbstractString) = "1.0"
f (generic function with 2 methods)

julia> g(a::A) = f(a.v)
g (generic function with 1 method)

julia> @code_warntype g(A(1.0))
Variables
  #self#::Core.Compiler.Const(g, false)
  a::A

Body::Union{Float64, String}
1 ─ %1 = Base.getproperty(a, :v)::Any
│   %2 = Main.f(%1)::Union{Float64, String}
└──      return %2
```